pkg/engine/combat/hit.go

```diff
@@ -27,22 +29,82 @@ func (mgr *Manager) performHit(hit *info.Hit) {
 	// * dots & element damage do not crit (unknown if also ByPureDamage?)
 	// * ByPureDamage = true means a "simplified" damage function (the break damage equation)
 
+	base := mgr.baseDamage(hit)*hit.HitRatio + hit.DamageValue
+	bonus := mgr.bonusDamage(hit)
+	crit := mgr.crit(hit)
+	total := mgr.totalDamage(hit, base, bonus, crit)
+	hpUpdate := mgr.shld.AbsorbDamage(hit.Defender.ID(), total)
+
+	mgr.attr.ModifyHPByAmount(hit.Defender.ID(), hit.Attacker.ID(), total, true)
+	mgr.attr.ModifyStance(hit.Defender.ID(), hit.Attacker.ID(), hit.StanceDamage*hit.HitRatio)
+	if mgr.target.IsCharacter(hit.Attacker.ID()) {
+		mgr.attr.ModifyEnergy(hit.Attacker.ID(), hit.EnergyGain*hit.HitRatio)
+	} else {
+		mgr.attr.ModifyEnergy(hit.Defender.ID(), hit.EnergyGain*hit.HitRatio)
+	}
+
 	mgr.event.HitEnd.Emit(event.HitEndEvent{
 		Attacker:         hit.Attacker.ID(),
 		Defender:         hit.Defender.ID(),
 		AttackType:       hit.AttackType,
 		DamageType:       hit.DamageType,
-		BaseDamage:       0, // TODO
-		BonusDamage:      0, // TODO
-		TotalDamage:      0, // TODO
-		ShieldDamage:     0, // TODO
-		HPDamage:         0, // TODO
+		HPDamage:         hpUpdate,
+		BaseDamage:       base,
+		BonusDamage:      bonus,
```

Comment:
nit: should `BonusDamage` be the percentage like it is currently, or `(base * bonus) - base` to represent the additional damage added from the bonus (or something else entirely?)